sympy/solvers/ode/tests/test_single.py

```diff
@@ -2563,9 +2563,9 @@ def _get_examples_ode_sol_nth_linear_constant_coeff_homogeneous():
     'lin_const_coeff_hom_37': {
         'eq': f(x).diff(x, 10) - 2*f(x).diff(x, 6) + 2*f(x).diff(x, 5)
         + f(x).diff(x, 2) - 2*f(x).diff(x, 1) + f(x),
-        'sol': [Eq(f(x), (C1 + C2*x)*exp(x*r21) + (C10*sin(x*im(r24)) + C7*x*sin(x*im(r24)) + (
-        C8 + C9*x)*cos(x*im(r24)))*exp(x*re(r24)) + (C3*x*sin(x*im(r22)) + C6*sin(x*im(r22)
-        ) + (C4 + C5*x)*cos(x*im(r22)))*exp(x*re(r22)))],
+        'sol': [Eq(f(x), (C1 + C2*x)*exp(x*r21) + (-((C3 + C4*x)*sin(x*im(r22)))
+        + (C5 + C6*x)*cos(x*im(r22)))*exp(x*re(r22)) + (-((C7 + C8*x)*sin(x*im(r24)))
+        + (C10*x + C9)*cos(x*im(r24)))*exp(x*re(r24)))],
```

Comment:
Why has this changed?

Reply:
I think because of ordering of constants and simplification function is called inside solver earlier this was handled in `odesimp`.

Reply:
Can we not just let `odesimp` handle this?

Reply:
In odesimp there was explicit condition for the hints having `nth_linear_const_coeff`, with the help of global variable it was handling the simplification which we discussed in the meeting. But now that simplification is moved in the solver.

Reply:
I still don't understand why we can't just let `odesimp` do all the simplification. If it can't figure out which expressions to collect on then maybe odesimp needs to be improved.

Reply:
so If we keep that simplification part in `odesimp` then we also need to pass the list of collectterms to it so that it can handle.
If you can suggest some other way, it would be helpful.

Reply:
Why can't `odesimp` just figure out which terms to collect on itself?

Reply:
In this case the terms are roots of that equation with real and imaginary parts also. I will try to think if it can be managed

Reply:
I had a look at `odesimp` it solves for `func` basically and handles constants. I couldn't come up with a way to handle our case. if you can have a look and suggest please.

Reply:
Wherever the old code is calling collect with `collectterms` it should be possible to add some code that calculates which terms to collect with by checking for e.g. `exp`, `sin` or `cos` terms that involve `t`.

Reply:
Old code is calling collect on `x**i*exp(reroot*x)*sin(abs(imroot)*x)` where i is the multiplicity. so if we want odesimp to handle this then it will require `(i,imroot,reroot)` also.

Reply:
Why does it need them? Why can't it just do `expr.atoms(exp)`?

The `simpsol` function in `systems.py` doesn't need to be told exactly which terms to collect on so why does `odesimp`?

Reply:
```python
sol = expand_mul(sol)
arr=eq[0].rhs.atoms(exp)
arr2=eq[0].rhs.atoms(sin)
arr3=eq[0].rhs.atoms(cos)
for i in arr:
    sol = collect(sol,i)
for i in arr2:
    sol = collect(sol,i)
for i in arr3:
    sol = collect(sol,i)
sol = powsimp(sol)
return Eq(f(x), sol)
```
I tried this, it changes a lot of current expected solutions for these hints.

Reply:
If you can make it pass the same arguments to `collect` then it should give the same results.

Reply:
how do i pass that because old code does `sol = collect(sol, x**i*exp(reroot*x)*sin(abs(imroot)*x))`

Reply:
Can you give an example of input and expected output as well as the arguments that are passed to `collect`?

Reply:
```python
>>> eq=-x**4*sin(x - 1) + f(x) + Derivative(f(x), (x, 2))
[(0, 0, -1), (0, 0, 1)]  #(multiplicity, reroot, imroot)
sol = collect(sol, x**i*exp(reroot*x)*sin(abs(imroot)*x))
sol = collect(sol, x**i*exp(reroot*x)*cos(imroot*x))
sol = collect(sol, x**i*exp(reroot*x))
sol = powsimp(sol)
expected_sol = Eq(f(x), C1*sin(x) + C2*cos(x) - x**5*cos(x - 1)/10 + x**4*sin(x - 1)/4 + x**3*cos(x - 1)/2 - 3*x**2*sin(x - 1)/4 - 3*x*cos(x - 1)/4)
```
The code snippet which I shared [here](https://github.com/sympy/sympy/pull/21654#discussion_r658712291) gives `Eq(f(x), C1*sin(x) + C2*cos(x) + x**2*(x**2 - 3)*sin(x - 1)/4 + x*(-2*x**4 + 10*x**2 - 15)*cos(x - 1)/20)`

Reply:
This discussion would be a lot quicker if you would communicate more clearly.

What is `sol` in this example?

Reply:
sorry for the miscommunication.
Let's take the example `eq=-8*f(x) + 12*Derivative(f(x), x) - 6*Derivative(f(x), (x, 2)) + Derivative(f(x), (x, 3))` the root is 2. so the collectterms (which is an array of tupe having `(multiplicity, realroot, imaginaryPart)`) will be  `[(2, 2, 0), (1, 2, 0), (0, 2, 0)]`.
general sol returned will be `Eq(f(x), C1*exp(2*x) + C2*x*exp(2*x) + C3*x**2*exp(2*x))`.
After doing collect on roots the expected simplified solution is `Eq(f(x), (C1 + x*(C2 + C3*x))*exp(2*x))`.
If it is still not clear, can we have discuss this on meet?

Reply:
We can do that like:
```python
In [24]: C1, C2, C3 = symbols('C1:4')

In [25]: eq = Eq(f(x), C1*exp(2*x) + C2*x*exp(2*x) + C3*x**2*exp(2*x)).rhs

In [26]: eq
Out[26]: 
    2⋅x         2⋅x       2  2⋅x
C₁⋅ℯ    + C₂⋅x⋅ℯ    + C₃⋅x ⋅ℯ   

In [27]: eq.collect(eq.atoms(exp))
Out[27]: 
⎛                2⎞  2⋅x
⎝C₁ + C₂⋅x + C₃⋅x ⎠⋅ℯ 


In [28]: eq.collect(eq.atoms(exp), horner)
Out[28]: 
                      2⋅x
(C₁ + x⋅(C₂ + C₃⋅x))⋅ℯ 
```
Actually I prefer `Out[27]` though. I don't see why we would want to change it to `Out[28]`.

Reply:
Yes `out[27]` is achievable, but this was just one of the examples. I think I should commit and change the examples and then we can discuss further if the examples look simplified or not.